File: Week_03/G20190343020039/LeetCode_102_0039.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
  if (root === null) { return []; }
  let result = [];
  let queue = [];
  queue.push(root);
  while (queue.length > 0) {
    const levelSize = queue.length;
    const currentLevel = [];
    for (let i = 0; i < levelSize; i += 1) {
      let node = queue.shift();
      currentLevel.push(node.val);
      if (node.left) { queue.push(node.left); }
      if (node.right) { queue.push(node.right); }
    }
    result.push(currentLevel);
  }
  return result;
};

File: Week_06/G20190343020039/LeetCode_102_0039.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from collections import deque
class Solution:
  def levelOrder(self, root: TreeNode) -> List[List[int]]:
    if root == None:
      return []
    result = []
    queue = deque([root,])
    while len(queue) > 0:
      levelSize = len(queue)
      currentLevel = []
      for i in range(levelSize):
        node = queue.popleft()
        currentLevel.append(node.val)
        if node.left:
          queue.append(node.left)
        if node.right:
          queue.append(node.right)
      result.append(currentLevel)
    return result

